choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectory.cpp

```diff
@@ -0,0 +1,103 @@
+#include "choreo/lib/ChoreoTrajectory.h"
+
+#include <wpi/json.h>
+
+using namespace choreolib;
+
+ChoreoTrajectory::ChoreoTrajectory(
+		const std::vector<ChoreoTrajectoryState> &states) : samples(states) {
+}
+
+ChoreoTrajectoryState ChoreoTrajectory::SampleInternal(
+		units::second_t timestamp) {
+	if (timestamp < samples[0].timestamp) {
+		return samples[0];
+	}
+	if (timestamp > GetTotalTime()) {
+		return samples[samples.size() - 1];
+	}
+
+	std::size_t low = 0;
+	std::size_t high = samples.size() - 1;
+
+	while (low != high) {
+		std::size_t mid = (low + high) / 2;
+		if (samples[mid].timestamp < timestamp) {
+			low = mid + 1;
+		} else {
+			high = mid;
+		}
+	}
+
+	if (low == 0) {
+		return samples[low];
+	}
+
+	ChoreoTrajectoryState behindState = samples[low - 1];
+	ChoreoTrajectoryState aheadState = samples[low];
+
+	if ((aheadState.timestamp - behindState.timestamp)
+			< units::second_t { 1e-6 }) {
+		return aheadState;
+	}
+
+	return behindState.Interpolate(aheadState, timestamp);
+}
+
+ChoreoTrajectoryState ChoreoTrajectory::Sample(units::second_t timestamp) {
+	return Sample(timestamp, false);
```

Comment:
You can remove this overload and use a default argument on the two-arg overload.